### ISSUE TITLE
Requesting VSCode support

### ISSUE THREAD

**Issue by palaniraja** [utk-se/CodeRibbon#80]
Wondering if vscode can be supported. I know there are some difference in extension arch between atom and vscode but would appreciate if this can be ported for vscode

**Comment by robobenklein** [comment]
Since this is absolutely the most requested "feature" I'm open to using this issue to track it until a separate repo is made for a VSC implementation.

Right now this implementation is tied fairly specifically to Atom's structure, in fact Patches are just subclasses of Panes and the Ribbon is a subclass of PaneAxis, which is what allows us to maintain compatibility with the rest of the Atom ecosystem.

AFAIK custom editors in VSCode do not allow nearly the same level of access to the internals of the editor, things that this package depends on include access to the DOM, overriding the entire workspace view, and stealing the editor's class prototypes from the existing Atom classes.

If there are people willing to help overcome the technical limitations I'm facing when trying to do the same in VSC then I welcome anyone with ideas to start a github discussion (just enabled the feature for this repo) and I'll try my best to use all the information I can get.

**Comment by colelawrence** [comment]
@robobenklein, I've opened [a discussion to share research findings on how this UX could be incrementally approached in VS Code](https://github.com/utk-se/CodeRibbon/discussions/84).

If anyone has good ideas on how they might tweak VS Code's source to start enabling this functionality, I'd love to collect your ideas, there!

**Comment by Jacke** [comment]
We need this. How much this could cost? Such as a minimum amount of donation? I've got sponsors

**Comment by robobenklein** [comment]
I'm currently funded as a PhD student so this summer I should have much more time to work on it without classes.

Even just one more dev familiar with VScode plugin development I could ping with random questions could help me make progress faster. (So much of the Atom and VScode internal APIs are difficult to search for due to search term collision with other software dev...)

If you do have a company / organization interested in backing the project I'd also love to get in touch to get more feedback from devs using it, since a big part of the research is to compare CodeRibbon vs Patchworks vs Split-View Editors. If you have existing VScode users willing to test out some alpha/beta quality builds that would certainly accelerate development.

**Comment by jtagcat** [comment]
jfyi I'm up for giving feedback on usage, and willing to test alpha versions.

As far as money goes, if @robobenklein wishes not to get funded, funding a VSCode dev to focus on the plugin/integration would be in scope.

**Comment by zhanba** [comment]
I have some VSCode extension development experience, if you need help, you can contact me.

**Comment by robobenklein** [comment]
I've created a new repository for VSCodeRibbon, though it is still a long ways off from being usable. I have moved a discussion there already. ( https://github.com/utk-se/VSCodeRibbon/discussions/5 )

I am still deciding on how to approach writing the changesets needed, but for now the plan is simply making commits over the current open source repo in a fork and using the VSCodium build tooling. Merging in upstream updates might be more difficult but at least it's easier to develop and won't break like a patchset might when microsoft pushes new versions.

As an aside, I've also enabled GitHub sponsors for those who wish to support the endeavor financially, which lets me spend more personal / non-research time on the project as well.

Although progress should be faster than the Atom implementation, CodeRibbon is still the result of more than two years of work and usage feedback. Since it'll take awhile I'm planning on posting some visual updates similar to the Visual History page on this repo's wiki. If there's enough interest I may even start a development blog series on the site, but for now real work on VSCodeRibbon comes first.

I'll eventually close this issue here, feel free to open issues or discussions on the new repo for questions or suggestions.

**Comment by robobenklein** [comment]
I feel bad for not updating very frequently, but *this semester* it's looking like launch time... definitely gotta have it ready for real work before [atom's sunset](https://github.blog/2022-06-08-sunsetting-atom/) at least.

Quick gif to make up for the notification:
![GIFrecord_2022-04-13_103136](https://user-images.githubusercontent.com/5423266/188063553-1b430fd0-271d-4fa2-b3ba-8762a503ef32.gif)

**Comment by palaniraja** [comment]
No worries, _school_ first. You have no obligation to complete an opensource project's feature request.

Thanks for keep working on it though, update looks good.

Good luck with your studies

**Comment by MartinBernstorff** [comment]
This would be **so** sweet! Is the project currently in a state where it can use help? :-)

**Comment by robobenklein** [comment]
Absolutely, I welcome PRs, and even just some feedback on which features should be prioritized is helpful!
(While it does technically *work*, it's lacking a lot of functionality that I would desire from a daily driver, especially drag-and-drop patch arrangement, which is why I'm still using an old version of Atom+CodeRibbon right now.)

You can try it: https://github.com/CodeRibbon/CodeRibbon-Theia

Feel free to open an issue for anything actionable and a discussion for anything else! (Even "I prefer other editor for reasons... / I hate CodeRibbon because ..." can be useful.)

Another progress gif (swapping columns):
![GIFrecord_2022-11-28_094418](https://user-images.githubusercontent.com/5423266/224506197-c629c641-c3a3-465f-9b7d-b675bb69d10f.gif)
